eth/downloader/queue.go

```diff
@@ -351,91 +364,77 @@ func (q *queue) Schedule(headers []*types.Header, from uint64) []*types.Header {
 	return inserts
 }
 
-// GetHeadResult retrieves the first fetch result from the cache, or nil if it hasn't
-// been downloaded yet (or simply non existent).
-func (q *queue) GetHeadResult() *fetchResult {
-	q.lock.RLock()
-	defer q.lock.RUnlock()
-
-	// If there are no results pending, return nil
-	if len(q.resultCache) == 0 || q.resultCache[0] == nil {
-		return nil
-	}
-	// If the next result is still incomplete, return nil
-	if q.resultCache[0].Pending > 0 {
-		return nil
-	}
-	// If the next result is the fast sync pivot...
-	if q.mode == FastSync && q.resultCache[0].Header.Number.Uint64() == q.fastSyncPivot {
-		// If the pivot state trie is still being pulled, return nil
-		if len(q.stateTaskPool) > 0 {
-			return nil
-		}
-		if q.PendingNodeData() > 0 {
-			return nil
-		}
-		// If the state is done, but not enough post-pivot headers were verified, stall...
-		for i := 0; i < fsHeaderForceVerify; i++ {
-			if i+1 >= len(q.resultCache) || q.resultCache[i+1] == nil {
-				return nil
-			}
-		}
-	}
-	return q.resultCache[0]
-}
-
-// TakeResults retrieves and permanently removes a batch of fetch results from
+// WaitResults retrieves and permanently removes a batch of fetch results from
 // the cache.
-func (q *queue) TakeResults() []*fetchResult {
+func (q *queue) WaitResults() (results []*fetchResult, closed bool) {
 	q.lock.Lock()
 	defer q.lock.Unlock()
 
-	// Accumulate all available results
-	results := []*fetchResult{}
-	for i, result := range q.resultCache {
-		// Stop if no more results are ready
-		if result == nil || result.Pending > 0 {
-			break
-		}
-		// The fast sync pivot block may only be processed after state fetch completes
-		if q.mode == FastSync && result.Header.Number.Uint64() == q.fastSyncPivot {
-			if len(q.stateTaskPool) > 0 {
-				break
+	// Wait for cached data.
+	getCacheLen := func() int {
+		for i, result := range q.resultCache {
+			if result == nil || result.Pending > 0 {
+				return i
 			}
-			if q.PendingNodeData() > 0 {
-				break
-			}
-			// Even is state fetch is done, ensure post-pivot headers passed verifications
-			safe := true
-			for j := 0; j < fsHeaderForceVerify; j++ {
-				if i+j+1 >= len(q.resultCache) || q.resultCache[i+j+1] == nil {
-					safe = false
+		}
+		return len(q.resultCache)
+	}
+	cacheLen := getCacheLen()
+	for cacheLen == 0 && !q.closed {
+		q.active.Wait()
+		cacheLen = getCacheLen()
+	}
+
+	for i := 0; i < cacheLen; i++ {
+		result := q.resultCache[i]
+		// The fast sync pivot block may only be processed after
+		// state fetch completes
```

Comment:
Could oyu move this comment down 2 lines to right above the `if len(q.stateTaskPool) > 0 {`? I think that's where it makes the most sense and in the current code structure it's a bit out of scope here (it wasn't at the best place in the original code either).